envs/quadrotors/quad_multi_model.py

```diff
@@ -13,25 +14,42 @@ def __init__(self, cfg, obs_space, timing, self_obs_dim=18, neighbor_obs_dim=6,
         self.self_obs_dim = self_obs_dim
         self.neighbor_obs_dim = neighbor_obs_dim
         self.neighbor_hidden_size = neighbor_hidden_size
+        self.use_spectral_norm = cfg.use_spectral_norm
 
         fc_encoder_layer = cfg.hidden_size
         # encode the current drone's observations
-        self.self_encoder = nn.Sequential(
-            nn.Linear(self.self_obs_dim, fc_encoder_layer),
-            nonlinearity(cfg),
-            nn.Linear(fc_encoder_layer, fc_encoder_layer),
-            nonlinearity(cfg)
-        )
+        if self.use_spectral_norm:
+            self.self_encoder = nn.Sequential(
+                spectral_norm(nn.Linear(self.self_obs_dim, fc_encoder_layer)),
+                nonlinearity(cfg),
+                spectral_norm(nn.Linear(fc_encoder_layer, fc_encoder_layer)),
+                nonlinearity(cfg)
+            )
+        else:
+            self.self_encoder = nn.Sequential(
+                nn.Linear(self.self_obs_dim, fc_encoder_layer),
+                nonlinearity(cfg),
+                nn.Linear(fc_encoder_layer, fc_encoder_layer),
+                nonlinearity(cfg)
+            )
         # encode the neighboring drone's observations
-        self.neighbor_encoder = nn.Sequential(
-            nn.Linear(self.neighbor_obs_dim, self.neighbor_hidden_size),
-            nonlinearity(cfg),
-        )
+        if self.use_spectral_norm:
+            self.neighbor_encoder = nn.Sequential(
+                spectral_norm(nn.Linear(self.neighbor_obs_dim, self.neighbor_hidden_size)),
```

Comment:
Unrelated to spectral norm, but do you think it makes sense to do 2 fc layers before feature aggregation?
Intuitively, just a single linear transformation + tanh might not be enough to get the best features.

I'd definitely try 2 fc layers before mean() and concat()